### DISCUSSION TITLE
Apostrophe 3.10.0: The last release of 2021! 🍾

### DISCUSSION THREAD

**Discussion by agilbert** [apostrophecms/apostrophe#3599] [Release Notes]
Hi everyone,

Today we are sharing our last release of 2021 with you before taking a pause to enjoy some holiday time with family and friends.

This release includes some improvements to Apostrophe core along with a few additional beta modules that we’ve wrapped up and slid under the 🎄:

- [Security Headers](https://github.com/apostrophecms/security-headers) is a new optional module that sends HTTP security headers that provide extra defense against CSRF attacks and are expected by various security scanners.
- [Anchors](https://github.com/apostrophecms/anchors) adds a wrapping element with an anchor linking target around all widgets, enabling the ability to manage links that jump to anchors on a page.
- [Redis Cache](https://github.com/apostrophecms/cache-redis) enhances the core caching module (`@apostrophecms/cache`) to use Redis rather than MongoDB.

These new modules are all ports from A2 that have been improved with A3 capabilities in mind. Remember, if you are interested in porting your own modules and projects from A2 to A3, [we have new tools and documentation available for that and are interested in your feedback](https://github.com/apostrophecms/code-upgrader).

It’s been an amazing year of releases for Apostrophe - certainly one of our biggest ever. Looking ahead, we will only be **increasing** in velocity with new features and updates in 2022. For now, we hope you are all closing out 2021 safe and healthy with some exciting plans for new Apostrophe projects in the new year!

🥂

### Fixes

* `slug` type fields can now have an empty string or `null` as their `def` value without the string `'none'` populating automatically.
* The `underline` feature works properly in tiptap toolbar configuration.
* Required checkbox fields now properly prevent editor submission when empty.
* Pins `vue-click-outside-element` to a version that does not attempt to use `eval` in its distribution build, which is incompatible with a strict Content Security Policy.

### Adds

* Adds a `last` option to fields. Setting `last: true` on a field puts that field at the end of the field's widget order. If more than one field has that option active the true last item will depend on general field registration order. If the field is ordered with the `fields.order` array or field group ordering, those specified orders will take precedence.

### Changes

* Adds deprecation notes to the widget class methods `getWidgetWrapperClasses` and `getWidgetClasses` from A2.
* Adds a deprecation note to the `reorganize` query builder for the next major version.
* Uses the runtime build of Vue. This has major performance and bundle size benefits, however it does require changes to Apostrophe admin UI apps that use a `template` property (components should require no changes, just apps require an update). These apps must now use a `render` function instead. Since custom admin UI apps are not yet a documented feature we do not regard this as a bc break.
* Compatible with the `@apostrophecms/security-headers` module, which supports a strict `Content-Security-Policy`.
* Adds a deprecation note to the `addLateCriteria` query builder.
* Updates the `toCount` doc type query method to use Math.ceil rather than Math.floor plus an additional step.

## Apostrophe 3.x modules

[anchors 1.0.0-beta](https://github.com/apostrophecms/anchors)

* Initial module release supporting widget wrapping for anchor linking.

[cache-redis 1.0.0-beta](https://github.com/apostrophecms/cache-redis)

* The initial beta release, including Redis client support for the standard Apostrophe cache API.

[security-headers 1.0.0-beta](https://github.com/apostrophecms/security-headers)

- This module sends the modern HTTP security headers that provide extra defense against CSRF attacks and are expected by various security scanners.

[seo 1.1.1](https://github.com/apostrophecms/seo)

- Compatible with the `@apostrophecms/security-headers` module.

**Comment by myovchev** [comment]
Extremely useful features and a perfect Christmas gift for the community :)
Cheers all.
